File: src/deadlock_build_sync/offline/refresh.py
"""Refresh production evidence without model training or comparison reports."""

from __future__ import annotations

import argparse
import os
from dataclasses import replace
from datetime import UTC, datetime
from pathlib import Path

from deadlock_build_sync.guide_generator import GENERATOR_NAMES
from deadlock_build_sync.value_validation import integer, require_object_dict

from .api import capture_sources, read_json, write_json
from .beam_snapshot import beam_resume_record, require_beam_resume
from .config import Cohort, RunPaths, parse_timestamp
from .extract import extract_cohort
from .production_evidence import export_production_evidence
from .production_sources import _select_patch_at_timestamp


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--run-id")
    parser.add_argument("--since")
    parser.add_argument("--as-of")
    parser.add_argument("--min-rank", type=int, default=71)
    parser.add_argument("--max-rank", type=int, default=115)
    parser.add_argument("--rank-expansion", choices=("auto", "off"), default="auto")
    parser.add_argument("--workers", type=int, default=8)
    parser.add_argument("--resume", action="store_true")
    parser.add_argument("--generator", choices=GENERATOR_NAMES, default="current")
    args = parser.parse_args(argv)
    if args.workers < 1:
        parser.error("--workers must be at least 1")
    if args.resume and not args.run_id:
        parser.error("--resume requires --run-id")
    state = Path(os.environ.get("XDG_STATE_HOME", Path.home() / ".local/state"))
    paths = RunPaths.create(state / "deadlock-build-sync/offline", args.run_id)
    if args.resume:
        _validate_resume_request(paths, args)
    else:
        _capture_run_sources(paths, args)
    document = export_production_evidence(
        paths,
        args.output,
        workers=args.workers,
        resume=args.resume,
        generator=args.generator,
    )
    print(f"Evidence: {args.output} ({document['artifact_id']})")
    print(f"Source and admission reports: {paths.run}")
    return 0


def _validate_resume_request(paths: RunPaths, args: argparse.Namespace) -> None:
    manifest = require_object_dict(read_json(paths.run / "manifest.json"))
    cohort = require_object_dict(manifest["cohort"])
    if manifest.get("generator", "current") != getattr(args, "generator", "current"):
        raise ValueError("Resume generator must match the source snapshot")
    if manifest.get("generator") == "beam":
        require_beam_resume(manifest)
    if (
        manifest.get("schema_version") != 2
        or manifest.get("production_method") != "eclat_leiden_pairwise"
        or manifest.get("test_usage") != "reserved"
        or not manifest.get("extraction")
        or not (paths.raw / "analysis.duckdb").is_file()
    ):
        raise ValueError("Resume requires a completed production source extraction")
    if (
        manifest.get("rank_expansion") != args.rank_expansion
        or cohort["minimum_badge"] != args.min_rank
        or cohort["maximum_badge"] != args.max_rank
    ):
        raise ValueError("Resume rank options must match the source snapshot")
    for name in ("since", "as_of"):
        requested = getattr(args, name)
        if requested and parse_timestamp(requested) != parse_timestamp(
            str(cohort[name])
        ):
            raise ValueError("Resume time options must match the source snapshot")


def _capture_run_sources(paths: RunPaths, args: argparse.Namespace) -> None:
    if (paths.run / "manifest.json").exists():
        raise ValueError("Use a new --run-id to preserve the previous source snapshot")
    cohort = Cohort(
        minimum_badge=args.min_rank,
        maximum_badge=args.max_rank,
        as_of=parse_timestamp(args.as_of)
        or datetime.now(tz=UTC).replace(microsecond=0),
    )
    sources = capture_sources(paths)
    patch = _select_patch_at_timestamp(paths, cohort.resolved_as_of())
    patch_start = datetime.fromtimestamp(integer(patch["start_timestamp"]), UTC)
    requested = parse_timestamp(args.since) or cohort.since
    cohort = replace(cohort, since=max(requested, patch_start))
    cohort.validate()
    manifest = {
        "schema_version": 2,
        "rank_expansion": args.rank_expansion,
        "cohort": cohort.as_dict(),
        "sources": sources,
        "production_method": "eclat_leiden_pairwise",
        "test_usage": "reserved",
    }
    if getattr(args, "generator", "current") != "current":
        manifest["generator"] = args.generator
        manifest["beam_resume"] = beam_resume_record()
    write_json(paths.run / "manifest.json", manifest)
    manifest["extraction"] = extract_cohort(
        paths, cohort, rank_expansion=args.rank_expansion
    )
    write_json(paths.run / "manifest.json", manifest)

File: crates/deadlock-analysis/src/refresh.rs
use std::fs::OpenOptions;
use std::path::PathBuf;

use deadlock_data::{Error, Result, atomic_write_json, read_json, text, trace_operation};
use deadlock_guides::{BuildGenerator, CURRENT_METHOD_VERSION, beam_generator_record};
use deadlock_input::parse_patch_feed;
use serde_json::json;

use crate::config::{
    Cohort, RANK_RESET_AT, RefreshRequest, RunPaths, cohort_ranks, implementation_record,
    parse_timestamp,
};
use crate::extraction::extract_cohort;
use crate::production_evidence::export_evidence;
use crate::sources::capture_sources;

#[derive(Clone, Debug)]
pub struct RefreshResult {
    pub artifact_id: String,
    pub output: PathBuf,
    pub run_directory: PathBuf,
}

/// # Errors
/// Returns an error when source capture, snapshot identity, analysis, or complete artifact validation fails.
pub fn refresh_evidence(request: &RefreshRequest) -> Result<RefreshResult> {
    if request.workers == 0 || (request.resume && request.run_id.is_none()) {
        return Err(Error::new(
            "Refresh requires positive workers. Resume requires --run-id.",
        ));
    }
    request.ranks.validate()?;
    let paths = RunPaths::create(&request.root, request.run_id.as_deref())?;
    let lock = OpenOptions::new()
        .create(true)
        .truncate(false)
        .read(true)
        .write(true)
        .open(paths.run.join(".refresh.lock"))?;
    lock.try_lock()
        .map_err(|error| Error::new(format!("Analysis run is locked: {error}")))?;
    if request.resume {
        validate_resume(&paths, request)?;
    } else {
        capture_run(&paths, request)?;
    }
    let document = export_evidence(
        &paths,
        &request.output,
        request.workers,
        request.resume,
        request.generator,
        &request.api_base_url,
    )?;
    Ok(RefreshResult {
        artifact_id: text(&document, "artifact_id")?.into(),
        output: request.output.clone(),
        run_directory: paths.run,
    })
}

fn validate_resume(paths: &RunPaths, request: &RefreshRequest) -> Result<()> {
    let manifest = read_json(&paths.run.join("manifest.json"))?;
    let generator = manifest["generator"].as_str().unwrap_or("current");
    if manifest["schema_version"] != 2
        || manifest["production_method"] != "eclat_leiden_pairwise"
        || manifest["test_usage"] != "reserved"
        || manifest["extraction"]
            .as_object()
            .is_none_or(serde_json::Map::is_empty)
        || !paths.raw.join("analysis.duckdb").is_file()
    {
        return Err(Error::new(
            "Resume requires a completed production source extraction",
        ));
    }
    if manifest["method_version"] != CURRENT_METHOD_VERSION
        || manifest["implementation"] != implementation_record()
    {
        return Err(Error::new(
            "Source extraction implementation differs. Start a new --run-id.",
        ));
    }
    if generator != request.generator.as_str()
        || manifest["rank_expansion"] != serde_json::to_value(request.rank_expansion)?
        || cohort_ranks(&manifest["cohort"])? != request.ranks
    {
        return Err(Error::new(
            "Resume generator and ranks must match the source snapshot",
        ));
    }
    for (name, requested) in [("since", request.since), ("as_of", request.as_of)] {
        if let Some(requested) = requested
            && requested != parse_timestamp(text(&manifest["cohort"], name)?)?
        {
            return Err(Error::new(
                "Resume timestamps must match the source snapshot",
            ));
        }
    }
    Ok(())
}

fn capture_run(paths: &RunPaths, request: &RefreshRequest) -> Result<()> {
    let manifest_path = paths.run.join("manifest.json");
    if manifest_path.exists() {
        return Err(Error::new(
            "Use a new --run-id to preserve the previous source snapshot",
        ));
    }
    let as_of = request
        .as_of
        .unwrap_or_else(|| chrono::Utc::now().timestamp());
    let sources = trace_operation("analysis.capture_sources", Some("capture_sources"), || {
        capture_sources(paths, &request.api_base_url)
    })?;
    let patch = parse_patch_feed(&read_json(&paths.raw.join("patches.json"))?, Some(as_of))?;
    let since = request
        .since
        .unwrap_or(parse_timestamp(RANK_RESET_AT)?)
        .max(patch.start_timestamp);
    let cohort = Cohort {
        ranks: request.ranks,
        since,
        as_of,
    };
    cohort.validate()?;
    let mut manifest = json!({"schema_version":2,"rank_expansion":request.rank_expansion,"cohort":cohort.to_document()?,"sources":sources,
        "production_method":"eclat_leiden_pairwise","method_version":CURRENT_METHOD_VERSION,"implementation":implementation_record(),"test_usage":"reserved"});
    if request.generator == BuildGenerator::Beam {
        manifest["generator"] = "beam".into();
        manifest["beam_resume"] =
            json!({"generator":beam_generator_record()?,"implementation":implementation_record()});
    }
    atomic_write_json(&manifest_path, &manifest)?;
    manifest["extraction"] = extract_cohort(paths, &cohort, request.rank_expansion)?;
    atomic_write_json(&manifest_path, &manifest)
}
